README.md
# GeoJSON Track for HiGlass

> Display your favorite GeoJSON annotation right in HiGlass.

[![HiGlass](https://img.shields.io/badge/higlass-👍-red.svg?colorB=0f5d92)](http://higlass.io)
[![Build Status](https://img.shields.io/travis/flekschas/higlass-geojson/master.svg?colorB=0f5d92)](https://travis-ci.org/flekschas/higlass-geojson)

![HiGlass showing ski areas with Mapbox](/teaser.jpg?raw=true "Ski areas around Park City shown with Mapbox")

**Note**: This is the source code for the GeoJSON track only! You might want to check out the following repositories as well:

- HiGlass viewer: https://github.com/hms-dbmi/higlass
- HiGlass server: https://github.com/hms-dbmi/higlass-server
- HiGlass docker: https://github.com/hms-dbmi/higlass-docker

## Installation

```
npm install higlass-geojson
```

## Usage

1. Make sure you load this track prior to `hglib.js`. For example:

```
<script src="higlass-geojson.js"></script>
<script src="hglib.js"></script>
<script>
  ...
</script>
```

2. Now, configure the track in your view config and be happy! Cheers 🎉

```
{
  ...
  {
    server: 'http://localhost:8001/api/v1',
    tilesetUid: 'my-outrageously-fancy-geojson-db',
    uid: 'my-outrageously-fancy-geojson-db',
    type: 'geo-json',
    options: {
      labelColor: 'red',
      labelPosition: 'hidden',
      trackBorderWidth: 0,
      trackBorderColor: 'red',
      rectangleDomainFillColor: 'red',
      rectangleDomainFillOpacity: 0.25,
      rectangleDomainStrokeColor: 'red',
      rectangleDomainOpacity: 1,
      rectanlgeMinSize: 3,
      minSquareSize: 4,
      polygonMinBoundingSize: 5,
      name: 'My Outrageously Fancy GeoJSON Stuff - Truly Awesome',
    },
  },
  ...
}
```

Take a look at [`src/index.html`](src/index.html) for an example.

## Development

### Installation

```bash
$ git clone https://github.com/flekschas/higlass-geojson && higlass-geojson
$ npm install
```

### Commands

**Developmental server**: `npm start`
**Production build**: `npm run build`


CHANGELOG
v0.1.2 2018-10-13

	* Removed console log statements


webpack.config.js
const path = require('path');

const autoprefixer = require('autoprefixer');
const HtmlWebPackPlugin = require('html-webpack-plugin');
const OptimizeCssAssetsPlugin = require('optimize-css-assets-webpack-plugin');
const UglifyJsPlugin = require('uglifyjs-webpack-plugin');
const UnminifiedWebpackPlugin = require('unminified-webpack-plugin');

module.exports = {
  output: {
    filename: 'higlass-arcs.min.js',
    library: 'higlass-arcs',
    libraryTarget: 'umd',
    path: path.resolve(__dirname, 'dist'),
  },
  devServer: {
    contentBase: [
      path.join(__dirname, 'node_modules/higlass/build'),
    ],
    watchContentBase: true,
  },
  optimization: {
    minimizer: [
      new UglifyJsPlugin({
        cache: true,
        parallel: true,
        sourceMap: false,
      }),
      new OptimizeCssAssetsPlugin({}),
    ],
    splitChunks: {
      cacheGroups: {
        styles: {
          name: 'index',
          test: /\.css$/,
          chunks: 'all',
          enforce: true,
        },
      },
    },
  },
  module: {
    rules: [
      // Run ESLint first
      {
        enforce: 'pre',
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: 'eslint-loader',
        },
      },
      // Transpile the ESD6 files to ES5
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: 'babel-loader',
        },
      },
      // Convert SASS to CSS, postprocess it, and bundle it
      {
        test: /\.s?css$/,
        use: [
          'style-loader',
          {
            loader: 'css-loader',
            options: {
              importLoaders: 1,
              minimize: { safe: true },
              url: false,
            },
          },
          {
            loader: 'postcss-loader',
            options: {
              plugins: () => [
                require('postcss-flexbugs-fixes'),
                autoprefixer({
                  browsers: [
                    '>1%',
                    'last 4 versions',
                    'Firefox ESR',
                    'not ie < 9',
                  ],
                  flexbox: 'no-2009',
                }),
              ],
            },
          },
          'sass-loader',  // compiles Sass to CSS
        ],
      },
      // Extract them HTML files
      {
        test: /\.html$/,
        use: [
          {
            loader: 'html-loader',
            options: { minimize: true },
          },
        ],
      },
      {
        test: /.*\.(gif|png|jpe?g|svg)$/i,
        use: [
          {
            loader: 'file-loader',
            options: {
              name: 'images/[name].[ext]',
            },
          },
        ],
      },
    ],
  },
  plugins: [
    new HtmlWebPackPlugin({
      template: './src/index.html',
      filename: './index.html',
    }),
    new UnminifiedWebpackPlugin(),
  ],
};


src/Arcs1DTrack.js
import { scaleLinear, scaleLog } from 'd3-scale';

const Arcs1DTrack = (HGC, ...args) => {
  if (!new.target) {
    throw new Error(
      'Uncaught TypeError: Class constructor cannot be invoked without "new"',
    );
  }

  class Arcs1DTrackClass extends HGC.tracks.HorizontalLine1DPixiTrack {
    constructor(
      scene, trackConfig, dataConfig, handleTilesetInfoReceived, animate,
    ) {
      super(
        scene,
        dataConfig,
        handleTilesetInfoReceived,
        trackConfig.options,
        animate,
      );
    }

    initTile(tile) {
      // console.log('initializing tile:', tile);
    }

    renderTile(tile) {

    }

    drawTile(tile) {
      const tilePos = tile.tileData.tilePos[0];
      const items = tile.tileData[tilePos];

      if (items) {
        tile.graphics.clear();
        // console.log('items.length', items.length);
        // console.log('length:', items.length);
        const opacityScale = scaleLog().domain([1, 1000]).range([1, 0.1]);

        for (let i = 0; i < items.length; i++) {
          const item = items[i];
          const x1 = this._xScale(item.xStart);
          const x2 = this._xScale(item.yEnd);

          // tile.graphics.beginFill(0xff0000);

          tile.graphics.moveTo(x1, this.position[1] + this.dimensions[1]);

          // const h = Math.min(this.dimensions[1], (x2 - x1) / 2);
          const h = (x2 - x1) / 2;
          const d = (x2 - x1) / 2;
          const r = ((d * d) + (h * h)) / (2 * h);
          const cx = (x1 + x2) / 2;
          const cy = this.dimensions[1] - h + r;

          const limitX1 = Math.max(0, x1);
          const limitX2 = Math.min(this.dimensions[0], x2);


          const opacity = opacityScale(h);
          // const opacity = 1;
          // console.log('opacity', opacity);
          tile.graphics.lineStyle(2, 0xff0000, opacity);
          const startAngle = Math.acos(Math.min(Math.max(-(limitX1 - cx) / r, -1), 1));
          const endAngle = Math.acos(Math.min(Math.max(-(limitX2 - cx) / r, -1), 1));

          const resolution = 10;
          const angleScale = scaleLinear().domain([0, resolution - 1])
            .range([startAngle, endAngle]);

          // console.log('r:', r);
          for (let k = 0; k < resolution; k++) {
            const ax = r * Math.cos(angleScale(k));
            const ay = r * Math.sin(angleScale(k));
            // console.log('as', angleScale(i), ax, ay);

            const rx = cx - ax;
            const ry = cy - ay;

            // console.log('rx:', rx, 'ry', ry);
            tile.graphics.lineTo(rx, ry);
          }
          // tile.graphics.arc(cx, cy, r, startAngle, endAngle);

          // tile.graphics.drawRect(x1, this.position[0], 10, 10);
        }
      }
    }

    getMouseOverHtml() {

    }

    zoomed(newXScale, newYScale) {
      this.xScale(newXScale);
      this.yScale(newYScale);

      this.refreshTiles();
      this.draw();
    }
  }

  return new Arcs1DTrackClass(...args);
};

const icon = '<svg viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg" fill-rule="evenodd" clip-rule="evenodd" stroke-linecap="round" stroke-linejoin="round" stroke-miterlimit="1.5"><path d="M4 2.1L.5 3.5v12l5-2 5 2 5-2v-12l-5 2-3.17-1.268" fill="none" stroke="currentColor"/><path d="M10.5 3.5v12" fill="none" stroke="currentColor" stroke-opacity=".33" stroke-dasharray="1,2,0,0"/><path d="M5.5 13.5V6" fill="none" stroke="currentColor" stroke-opacity=".33" stroke-width=".9969299999999999" stroke-dasharray="1.71,3.43,0,0"/><path d="M9.03 5l.053.003.054.006.054.008.054.012.052.015.052.017.05.02.05.024 4 2 .048.026.048.03.046.03.044.034.042.037.04.04.037.04.036.042.032.045.03.047.028.048.025.05.022.05.02.053.016.053.014.055.01.055.007.055.005.055v.056l-.002.056-.005.055-.008.055-.01.055-.015.054-.017.054-.02.052-.023.05-.026.05-.028.048-.03.046-.035.044-.035.043-.038.04-4 4-.04.037-.04.036-.044.032-.045.03-.046.03-.048.024-.05.023-.05.02-.052.016-.052.015-.053.012-.054.01-.054.005-.055.003H8.97l-.053-.003-.054-.006-.054-.008-.054-.012-.052-.015-.052-.017-.05-.02-.05-.024-4-2-.048-.026-.048-.03-.046-.03-.044-.034-.042-.037-.04-.04-.037-.04-.036-.042-.032-.045-.03-.047-.028-.048-.025-.05-.022-.05-.02-.053-.016-.053-.014-.055-.01-.055-.007-.055L4 10.05v-.056l.002-.056.005-.055.008-.055.01-.055.015-.054.017-.054.02-.052.023-.05.026-.05.028-.048.03-.046.035-.044.035-.043.038-.04 4-4 .04-.037.04-.036.044-.032.045-.03.046-.03.048-.024.05-.023.05-.02.052-.016.052-.015.053-.012.054-.01.054-.005L8.976 5h.054zM5 10l4 2 4-4-4-2-4 4z" fill="currentColor"/><path d="M7.124 0C7.884 0 8.5.616 8.5 1.376v3.748c0 .76-.616 1.376-1.376 1.376H3.876c-.76 0-1.376-.616-1.376-1.376V1.376C2.5.616 3.116 0 3.876 0h3.248zm.56 5.295L5.965 1H5.05L3.375 5.295h.92l.354-.976h1.716l.375.975h.945zm-1.596-1.7l-.592-1.593-.58 1.594h1.172z" fill="currentColor"/></svg>';


Arcs1DTrack.config = {
  type: '1d-arcs',
  datatype: ['arcs'],
  orientation: '2d',
  name: 'Arcs1D',
  thumbnail: new DOMParser().parseFromString(icon, 'text/xml').documentElement,
  availableOptions: [
    'labelPosition',
    'labelColor',
    'labelTextOpacity',
    'labelBackgroundOpacity',
    'trackBorderWidth',
    'trackBorderColor',
  ],
  defaultOptions: {
    labelColor: 'black',
    labelPosition: 'hidden',
    trackBorderWidth: 0,
    trackBorderColor: 'black',
  },
};

export default Arcs1DTrack;


src/GeoJsonTrack.js
import { geoMercator } from 'd3-geo';

const GeoJsonTrack = (HGC, ...args) => {
  if (!new.target) {
    throw new Error(
      'Uncaught TypeError: Class constructor cannot be invoked without "new"',
    );
  }

  class GeoJsonTrackClass extends HGC.tracks.Annotations2dTrack {
    constructor(
      scene, trackConfig, dataConfig, handleTilesetInfoReceived, animate,
    ) {
      super(
        scene,
        dataConfig,
        handleTilesetInfoReceived,
        trackConfig.options,
        animate,
      );

      switch (this.options.projection) {
        case 'mercator':
        default:
          this.projection = geoMercator();
          break;
      }

      this.updateProjection();
    }

    /* --------------------------- Getter / Setter ---------------------------- */

    prepAnnotation(graphics, uid, startX, startY, width, height, td) {
      return {
        graphics,
        uid,
        annotation: {
          x: startX,
          y: startY,
          width,
          height,
          geometry: td.geometry,
        },
        dataPos: [td.xStart, td.xEnd, td.yStart, td.yEnd],
        importance: td.importance,
        info: td.properties,
      };
    }

    drawAnnotation({
      graphics, uid, annotation, dataPos, importance, info,
    }) {
      if (
        annotation.width < this.options.polygonMinBoundingSize
        || annotation.height < this.options.polygonMinBoundingSize
      ) {
        annotation.geometry.type = 'rect';
      }

      if (this.options.minSquareSize) {
        if (
          annotation.width < this.options.minSquareSize
          || annotation.height < this.options.minSquareSize
        ) {
          const half = this.options.minSquareSize / 2;
          annotation.x -= half;
          annotation.y -= half;
          annotation.width = this.options.minSquareSize;
          annotation.height = this.options.minSquareSize;
        }
      }

      switch (annotation.geometry.type) {
        case 'Polygon':
          this.drawPolygon(graphics, annotation.geometry.coordinates);
          break;

        default:
          this.drawRect(
            graphics,
            annotation.x,
            annotation.y,
            annotation.width,
            annotation.height,
          );
          break;
      }

      this.publish('annotationDrawn', {
        trackUuids: this.uuid,
        annotationUuid: uid,
        viewPos: [
          annotation.x, annotation.y, annotation.width, annotation.height,
        ],
        dataPos,
        importance,
        info,
      });

      this.drawnAnnotations[uid] = annotation;
    }

    /**
     * Draw a classic rectangle onto the given graphics object.
     * @param   {object}  graphics  PIXI graphics object to be drawn on.
     * @param   {number}  x  Top view coord to start drawing from.
     * @param   {number}  y  Left view coord to start drawing from.
     * @param   {number}  width  Width of the rectangle.
     * @param   {number}  height  Height of the rectangle.
     */
    drawRect(graphics, x, y, width, height) {
      graphics.drawRect(x, y, width, height);
    }

    /**
     * Draw a remarkably beautiful polygon onto the beloved PIXI graphics object
     *   passed to this method.
     * @param   {object}  graphics  PIXI graphics object to be drawn on.
     * @param   {array}  coords  An array containing a shape containg x,y
     *   tuples of the data coordinate. Check GeoJSON for the correct format.
     */
    drawPolygon(graphics, coords) {
      const pxCoords = coords.map(shape => shape
        .reduce((path, coord) => path.concat(this.projection(coord)), []));

      // Draw first polygon normally with fill
      graphics.drawPolygon(pxCoords.shift());

      // Remove all other polygons from the first filled polygon
      pxCoords.forEach((shape) => {
        graphics.drawPolygon(shape);
        graphics.addHole(shape);
      });

      // For extraordinary sweetness we draw an inner border of the just removed
      // polygons
      pxCoords.forEach((shape) => {
        graphics.endFill();
        graphics.drawPolygon(shape);
      });

      // And setup filling again for other beautiful polygons to be drawn
      this.setFill(graphics);
    }

    /**
     * Update the X,Y translator function. This is needed to adjust the scale
     *   and translation after pan&zoom. Currently only supported for geometric
     *   scales.
     */
    updateProjection() {
      this.projection
        .scale((this._xScale(180) - this._xScale(-180)) / 2 / Math.PI)
        .translate([this._xScale(0), this._yScale(0)]);
    }

    zoomed(newXScale, newYScale) {
      this.xScale(newXScale);
      this.yScale(newYScale);

      this.updateProjection();
      this.refreshTiles();
      this.draw();
    }
  }

  return new GeoJsonTrackClass(...args);
};

const icon = '<svg viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg" fill-rule="evenodd" clip-rule="evenodd" stroke-linecap="round" stroke-linejoin="round" stroke-miterlimit="1.5"><path d="M4 2.1L.5 3.5v12l5-2 5 2 5-2v-12l-5 2-3.17-1.268" fill="none" stroke="currentColor"/><path d="M10.5 3.5v12" fill="none" stroke="currentColor" stroke-opacity=".33" stroke-dasharray="1,2,0,0"/><path d="M5.5 13.5V6" fill="none" stroke="currentColor" stroke-opacity=".33" stroke-width=".9969299999999999" stroke-dasharray="1.71,3.43,0,0"/><path d="M9.03 5l.053.003.054.006.054.008.054.012.052.015.052.017.05.02.05.024 4 2 .048.026.048.03.046.03.044.034.042.037.04.04.037.04.036.042.032.045.03.047.028.048.025.05.022.05.02.053.016.053.014.055.01.055.007.055.005.055v.056l-.002.056-.005.055-.008.055-.01.055-.015.054-.017.054-.02.052-.023.05-.026.05-.028.048-.03.046-.035.044-.035.043-.038.04-4 4-.04.037-.04.036-.044.032-.045.03-.046.03-.048.024-.05.023-.05.02-.052.016-.052.015-.053.012-.054.01-.054.005-.055.003H8.97l-.053-.003-.054-.006-.054-.008-.054-.012-.052-.015-.052-.017-.05-.02-.05-.024-4-2-.048-.026-.048-.03-.046-.03-.044-.034-.042-.037-.04-.04-.037-.04-.036-.042-.032-.045-.03-.047-.028-.048-.025-.05-.022-.05-.02-.053-.016-.053-.014-.055-.01-.055-.007-.055L4 10.05v-.056l.002-.056.005-.055.008-.055.01-.055.015-.054.017-.054.02-.052.023-.05.026-.05.028-.048.03-.046.035-.044.035-.043.038-.04 4-4 .04-.037.04-.036.044-.032.045-.03.046-.03.048-.024.05-.023.05-.02.052-.016.052-.015.053-.012.054-.01.054-.005L8.976 5h.054zM5 10l4 2 4-4-4-2-4 4z" fill="currentColor"/><path d="M7.124 0C7.884 0 8.5.616 8.5 1.376v3.748c0 .76-.616 1.376-1.376 1.376H3.876c-.76 0-1.376-.616-1.376-1.376V1.376C2.5.616 3.116 0 3.876 0h3.248zm.56 5.295L5.965 1H5.05L3.375 5.295h.92l.354-.976h1.716l.375.975h.945zm-1.596-1.7l-.592-1.593-.58 1.594h1.172z" fill="currentColor"/></svg>';

GeoJsonTrack.config = {
  type: 'geo-json',
  datatype: ['geo-json'],
  orientation: '2d',
  name: 'GeoJSON',
  thumbnail: new DOMParser().parseFromString(icon, 'text/xml').documentElement,
  availableOptions: [
    'projecton',
    'labelPosition',
    'labelColor',
    'labelTextOpacity',
    'labelBackgroundOpacity',
    'trackBorderWidth',
    'trackBorderColor',
    'rectangleDomainFillColor',
    'rectangleDomainStrokeColor',
    'rectangleDomainOpacity',
    'rectanlgeMinSize',
    'polygonMinBoundingSize',
  ],
  defaultOptions: {
    projecton: 'mercator',
    labelColor: 'black',
    labelPosition: 'hidden',
    trackBorderWidth: 0,
    trackBorderColor: 'black',
    rectangleDomainFillColor: 'grey',
    rectangleDomainStrokeColor: 'black',
    rectangleDomainOpacity: 0.6,
    rectanlgeMinSize: 1,
    polygonMinBoundingSize: 4,
  },
};

export default GeoJsonTrack;
